### PR TITLE
fix: harden secret detection in exempt tokens

### PR DESCRIPTION
**Pull request by seratch**
## Summary

This PR supersedes #87 while preserving its two original goals:

- detect provider-specific credentials embedded in tokens that would otherwise
  be exempt as URLs or allowed file paths;
- normalize only a genuine leading `www.` host label in the URL guardrail.

The Secret Keys guardrail now establishes structural URL and file components
before decoding them. Each component is percent-decoded at most once using
strict UTF-8 handling, then evaluated independently. This prevents unrelated
URL components, padding, or allowed file extensions from supplying the length
or diversity needed to classify a short value as a secret.

Only provider-specific forms can lift an existing URL or file exemption:

- `sk-` and `sk_`
- `ghp_`
- Slack `xoxb-` and `xoxp-`
- `SG.`
- `hf_`
- AWS access key IDs matching `AKIA` plus exactly 16 uppercase letters or
  digits

Generic prefixes such as `key-`, `api-`, `apikey-`, `token-`, and `secret-`
retain their existing standalone behavior but do not cause ordinary
high-entropy URL slugs to lose their exemption.

The URL guardrail replaces unanchored `www.` replacement with removal of
exactly one leading `www.` label at the existing normalization points.

No public API, configuration, dependency, or `custom_regex` behavior changes
are introduced.

## Relationship to #87

PR #87 introduced the component-local, decode-once approach for embedded
provider credentials and the leading-`www.` normalization fix. This PR carries
those changes forward and incorporates additional review-driven corrections.

Compared with the current #87 head (`bb4a6eb`), this version additionally:

- uses Unicode XID continuation semantics for prefix boundaries, preventing
  Unicode identifier characters such as middle dots and combining characters
  from manufacturing a secret boundary;
- preserves dotted provider prefixes such as `SG.` across parsed host labels
  and malformed-URL fallback parsing without joining ordinary non-dotted
  prefixes across labels;
- handles nested presentation wrappers and sentence punctuation adjacent to
  confirmed closing delimiters without treating the wrapper as credential
  data;
- centralizes percent-escape validation while retaining strict one-pass
  decoding;
- adds focused regression and property coverage for these boundary cases;
- brings all touched code and tests within the repository's 100-character
  limit.

## Security properties

The regression coverage verifies that:

- the reported URL-query and allowed-extension bypasses are detected;
- literal, partially encoded, mixed-case encoded, and fully encoded
  provider-specific prefixes classify consistently after one decoding pass;
- double-encoded prefixes are not decoded recursively;
- malformed escapes and invalid UTF-8 remain raw and cannot create synthetic
  boundaries;
- URL components are scored independently;
- query punctuation and encoded separators inside a candidate remain candidate
  data;
- Unicode XID continuations do not create prefix boundaries, while genuine
  punctuation can;
- valid AWS access key IDs are detected without flagging ordinary `AKIA`
  strings or longer identifier continuations;
- generic natural-language URL slugs remain exempt;
- parsed and malformed URLs handle `SG.` consistently;
- nested presentation punctuation does not complete a short candidate;
- terminal URI punctuation remains credential data when it is not part of a
  confirmed presentation wrapper;
- findings continue to report the original whitespace token once.

### CODE REVIEW

File: src/guardrails/checks/text/secret_keys.py [hunk 1]
```diff
@@ -259,22 +279,481 @@ def _contains_allowed_pattern(text: str) -> bool:
...
+    """Yield component-local views from an opaque URL fragment.
+
+    Args:
+        fragment: Raw fragment text.
+
+    Yields:
+        Query-style fields or slash-delimited fragment components.
+    """
+    if "=" in fragment or "&" in fragment:
+        yield from _iter_query_components(fragment)
```
[the remark this answers is not in the document]
Comment (seratch) [reply src/guardrails/checks/text/secret_keys.py]: @codex review

File: src/guardrails/checks/text/secret_keys.py [hunk 2]
```diff
@@ -259,22 +279,487 @@ def _contains_allowed_pattern(text: str) -> bool:
...
+def _iter_fallback_components(text: str) -> Iterator[str]:
+    """Yield conservative components when URL parsing fails.
+
+    Args:
+        text: Raw malformed URL text.
+
+    Yields:
+        Non-empty pieces separated by structural URL delimiters.
+    """
+    for component in re.split(r"[\\/?#&=@]", text):
```
[the remark this answers is not in the document]
Comment (seratch) [reply src/guardrails/checks/text/secret_keys.py]: @codex review

File: src/guardrails/checks/text/secret_keys.py [hunk 3]
```diff
@@ -259,22 +279,499 @@ def _contains_allowed_pattern(text: str) -> bool:
...
+    remainder = text[scheme_match.end() :] if scheme_match else text
+
+    route, fragment_separator, fragment = remainder.partition("#")
+    authority_and_path, query_separator, query = route.partition("?")
+    authority, path_separator, path = authority_and_path.partition("/")
+
+    yield from _iter_netloc_components(authority)
+    if path_separator:
+        for component in _iter_path_components(path, strip_final_extension=True):
+            yield from _iter_dot_delimited_components(component)
```
[the remark this answers is not in the document]
Comment (seratch) [reply src/guardrails/checks/text/secret_keys.py]: @codex review

File: src/guardrails/checks/text/secret_keys.py [hunk 4]
```diff
@@ -259,22 +279,507 @@ def _contains_allowed_pattern(text: str) -> bool:
...
+        yield from _iter_path_components(text, strip_final_extension=True)
+        return
+
+    first_start = scheme_matches[0].start()
+    if first_start:
+        yield from _iter_fallback_components(text[:first_start])
+
+    for index, match in enumerate(scheme_matches):
+        end = scheme_matches[index + 1].start() if index + 1 < len(scheme_matches) else len(text)
+        leading_context = text[: match.start()]
```
[the remark this answers is not in the document]
Comment (seratch) [reply src/guardrails/checks/text/secret_keys.py]: @codex review

File: src/guardrails/checks/text/secret_keys.py [hunk 5]
```diff
@@ -259,22 +279,543 @@ def _contains_allowed_pattern(text: str) -> bool:
...
+            text = f"{text[:first_start]}{trimmed_tail}"
+            scheme_matches = list(_URL_SCHEME_RE.finditer(text))
+
+    if first_start:
+        yield from _iter_fallback_components(text[:first_start])
+
+    for index, match in enumerate(scheme_matches):
+        end = scheme_matches[index + 1].start() if index + 1 < len(scheme_matches) else len(text)
+        leading_context = _presentation_openers_before(text, match.start())
+        url_text = _trim_url_presentation_suffix(text[match.start() : end], leading_context)
```
[the remark this answers is not in the document]
Comment (seratch) [reply src/guardrails/checks/text/secret_keys.py]: @codex review

File: src/guardrails/checks/text/secret_keys.py [hunk 6]
```diff
@@ -259,22 +283,567 @@ def _contains_allowed_pattern(text: str) -> bool:
...
+            if remainder.startswith(":"):
+                if len(remainder) > 1:
+                    yield remainder[1:]
+            elif remainder:
+                yield from _iter_dot_delimited_components(remainder)
+            return
+    else:
+        possible_host, port_separator, possible_port = host.rpartition(":")
+        if port_separator and possible_port.isdigit():
+            host = possible_host
```
[the remark this answers is not in the document]
Comment (seratch) [reply src/guardrails/checks/text/secret_keys.py]: @codex review
